Content.Shared/Doors/Systems/TurnstileSystem.cs

```diff
@@ -0,0 +1,109 @@
+    [Dependency] private readonly SharedAudioSystem _audio = default!;
+    [Dependency] private readonly EntityWhitelistSystem _entityWhitelist = default!;
+    [Dependency] private readonly PullingSystem _pulling = default!;
+    [Dependency] private readonly SharedPopupSystem _popup = default!;
+
+    /// <inheritdoc/>
+    public override void Initialize()
+    {
+        SubscribeLocalEvent<TurnstileComponent, PreventCollideEvent>(OnPreventCollide);
+        SubscribeLocalEvent<TurnstileComponent, EndCollideEvent>(OnEndCollide);
+    }
+
+    private void OnPreventCollide(Entity<TurnstileComponent> ent, ref PreventCollideEvent args)
+    {
+        if (args.Cancelled || !args.OurFixture.Hard)
+            return;
+
+        if (ent.Comp.CollideExceptions.Contains(args.OtherEntity))
+        {
+            // We need to add this in here too for chain pulls
+            if (_pulling.GetPulling(args.OtherEntity) is { } uid &&
+                ent.Comp.CollideExceptions.Add(uid))
+            {
+                Dirty(ent);
+            }
+
+            args.Cancelled = true;
+            return;
+        }
+
+        // unblockables go through for free.
+        if (_entityWhitelist.IsWhitelistFail(ent.Comp.ProcessWhitelist, args.OtherEntity))
+        {
+            args.Cancelled = true;
+            return;
+        }
+
+        if (_timing.CurTime < ent.Comp.NextPassTime)
+            return;
+
+        if (!_accessReader.IsAllowed(args.OtherEntity, ent))
+        {
+            _audio.PlayPredicted(ent.Comp.DenySound, ent, args.OtherEntity);
+            if (_net.IsClient)
+                _animationPlayer.Flick(ent, ent.Comp.DenyState, TurnstileVisualLayers.Base);
+            return;
+        }
+
+        var xform = Transform(ent);
+        var otherXform = Transform(args.OtherEntity);
+
+        var moveDir = (xform.LocalPosition - otherXform.LocalPosition).GetDir();
+        var faceDir = xform.LocalRotation.GetDir();
+
+        if (moveDir == faceDir)
+        {
+            ent.Comp.CollideExceptions.Add(args.OtherEntity);
+            if (_pulling.GetPulling(args.OtherEntity) is { } uid)
+                ent.Comp.CollideExceptions.Add(uid);
+
+            args.Cancelled = true;
+            ent.Comp.NextPassTime = _timing.CurTime + ent.Comp.PassDelay;
+            if (_net.IsClient)
+                _animationPlayer.Flick(ent, ent.Comp.SpinState, TurnstileVisualLayers.Base);
+            _audio.PlayPredicted(ent.Comp.TurnSound, ent, args.OtherEntity);
+            Dirty(ent);
+        }
+        else
+        {
+            if (_timing.CurTime >= ent.Comp.NextResistTime)
+            {
+                _popup.PopupPredicted(Loc.GetString("turnstile-component-popup-resist", ("turnstile", ent.Owner)), ent, args.OtherEntity);
+                ent.Comp.NextResistTime = _timing.CurTime + TimeSpan.FromSeconds(0.5);
+                Dirty(ent);
+            }
+        }
+    }
+
```

Comment:
Doing stuff in attempt events is really suss, also this is going to be spammed every single frame in collision which is the most expensive part of box2d which is not good.

I think you're better off just cancelling either the hard / non-hard collision as appropriate and just running the rest of the logic in startcollideevent so it also predicts properly when contacts get fully predicted and doesn't suddenly break in the next few months.